jupyter_scheduler/tests/test_handlers.py

```diff
@@ -1,13 +1,262 @@
+    with patch("jupyter_scheduler.scheduler.Scheduler.list_jobs") as mock_list_jobs:
+        mock_list_jobs.return_value = expected_jobs
+        response = await jp_fetch("scheduler", "jobs", method="GET", params=params)
+
+        mock_list_jobs.assert_called_once_with(expected_query)
+        assert response.code == 200
+        actual_job = json.loads(response.body)
+        actual_job = actual_job["jobs"][0]
+        expected_job = jobs_list["jobs"][0]
+        assert actual_job["input_uri"] == expected_job["input_uri"]
+        assert actual_job["output_prefix"] == expected_job["output_prefix"]
+        assert actual_job["runtime_environment_name"] == expected_job["runtime_environment_name"]
+        assert actual_job["job_id"] == expected_job["job_id"]
+        assert actual_job["output_uri"] == expected_job["output_uri"]
+        assert actual_job["url"] == expected_job["url"]
+
+
+async def test_patch_jobs_for_missing_status(jp_fetch):
+    with pytest.raises(HTTPClientError) as e:
+        job_id = "542e0fac-1274-4a78-8340-a850bdb559c8"
+        response = await jp_fetch("scheduler", "jobs", job_id, method="PATCH", body="{}")
+    assert expected_http_error(e, 500, "Field 'status' missing in request body")
+
+
+async def test_patch_jobs_for_stop_job(jp_fetch):
+    with patch("jupyter_scheduler.scheduler.Scheduler.stop_job") as mock_stop_job:
+        job_id = "542e0fac-1274-4a78-8340-a850bdb559c8"
+        response = await jp_fetch(
+            "scheduler", "jobs", job_id, method="PATCH", body=json.dumps({"status": "STOPPED"})
+        )
+
+        mock_stop_job.assert_called_once_with(job_id)
+        assert response.code == 204
+
+
+# TODO: Enable after fixing patch api
+@pytest.mark.skip
+async def test_patch_jobs_for_name_update(jp_fetch):
+    with patch("jupyter_scheduler.scheduler.Scheduler.stop_job") as mock_stop_job:
+        job_id = "542e0fac-1274-4a78-8340-a850bdb559c8"
+        response = await jp_fetch(
+            "scheduler", "jobs", job_id, method="PATCH", body=json.dumps({"name": "New job name"})
+        )
+
+        mock_stop_job.assert_called_once_with(job_id)
+        assert response.code == 204
+
+
+async def test_delete_job(jp_fetch):
+    with patch("jupyter_scheduler.scheduler.Scheduler.delete_job") as mock_delete_job:
+        job_id = "542e0fac-1274-4a78-8340-a850bdb559c8"
+        response = await jp_fetch("scheduler", "jobs", job_id, method="DELETE")
+
+        mock_delete_job.assert_called_once_with(job_id)
+        assert response.code == 204
+
+
+async def test_batch_delete(jp_fetch):
+    with patch("jupyter_scheduler.scheduler.Scheduler.delete_job") as mock_delete_job:
+        job_id = "542e0fac-1274-4a78-8340-a850bdb559c8"
+        response = await jp_fetch(
+            "scheduler", "batch", "jobs", method="DELETE", params={"job_id": job_id}
+        )
+
+        mock_delete_job.assert_called_once_with(job_id)
+        assert response.code == 204
+
```

Comment:
These tests (patch/delete) are not meaningful as they should be. You are only asserting that the response code is OK, and not verifying their intended side effect of updating/deleting jobs. Ideally, each of your existing assertions should be followed by a get request that actually verifies the corresponding job was updated/deleted.

Reply:
My thought process was adding testing in layers; these tests covering the API layer, only test the api portion making sure they call the correct scheduler method (mocking the scheduler methods here) with expected parameters. There are not a lot of exception cases here, but in future I expect some more tests to cover those scenarios.
I intend to write addition tests that will cover the scheduler portion of the code which will test actual persistence of jobs, but will add those in a separate test file.

Reply:
That's fine if you plan on adding more tests in the future. I do have a high-level comment: while I understand that your testing strategy conforms more strictly to proper "unit testing", I just don't see the value in separating the handler and business logic tests. IMO, it's just easier to write tests that cover both, while also giving you more confidence that your server actually works.

Reply:
@dlqqq Thanks for the suggestions. I agree there is value in integration tests that cover the full service function, but as we are evolving and changing things frequently right now, having unit tests keeps us more pliable at the moment. 
I am also following what some other jupyterlab extensions are doing. For example, here are the tests for jupyterlab-git extension.
https://github.com/jupyterlab/jupyterlab-git/blob/master/jupyterlab_git/tests/test_handlers.py